lib/vtls/schannel.c

```diff
@@ -1574,7 +1594,332 @@ CURLcode Curl_schannel_random(unsigned char *entropy, size_t length)
+        more_certs = 0;
+      }
+      else{
+        char *end_cert_ptr = strstr(begin_cert_ptr, END_CERT);
+        if(!end_cert_ptr) {
+          failf(data,
+                "schannel: CA file is not correctly formatted '%s'",
+                ca_file);
+          result = CURLE_SSL_CACERT_BADFILE;
+          more_certs = 0;
+        }
+        else{
+          CERT_BLOB cert_blob;
+          CERT_CONTEXT *cert_context = NULL;
+          BOOL add_cert_result = FALSE;
+          DWORD actual_content_type = 0;
+          DWORD cert_size = (DWORD)
+            ((end_cert_ptr + END_CERT_LEN) - begin_cert_ptr);
+
+          cert_blob.pbData = (BYTE *)begin_cert_ptr;
+          cert_blob.cbData = cert_size;
+          if(!CryptQueryObject(CERT_QUERY_OBJECT_BLOB,
+                                &cert_blob,
+                                CERT_QUERY_CONTENT_FLAG_CERT,
+                                CERT_QUERY_FORMAT_FLAG_ALL,
+                                0,
+                                NULL,
+                                &actual_content_type,
+                                NULL,
+                                NULL,
+                                NULL,
+                                &cert_context)) {
+
+            failf(data,
+                  "schannel: failed to extract certificate from CA file: %s",
+                  Curl_strerror(conn, GetLastError()));
+            result = CURLE_SSL_CACERT_BADFILE;
+            more_certs = 0;
+          }
+          else{
+            current_ca_file_ptr = current_ca_file_ptr + cert_size;
+
+            /* Sanity check that the cert_context object is the right type */
+            if(CERT_QUERY_CONTENT_CERT != actual_content_type) {
+              failf(data,
+                    "schannel: unexpected content type when extracting "
+                    "certificate: %d", actual_content_type);
+              result = CURLE_SSL_CACERT_BADFILE;
+              more_certs = 0;
+            }
+            else{
+              add_cert_result =
+                CertAddCertificateContextToStore(trust_store,
+                                                 cert_context,
+                                                 CERT_STORE_ADD_ALWAYS,
+                                                 NULL);
+              CertFreeCertificateContext(cert_context);
+              if(!add_cert_result) {
+                failf(data,
+                      "schannel: failed to add certificate from CA file "
+                      "to certificate store: %s",
+                      Curl_strerror(conn, GetLastError()));
+                result = CURLE_SSL_CACERT_BADFILE;
+                more_certs = 0;
+              }
+              else{
+                found_certs = 1;
+              }
+            }
+          }
+        }
+      }
+    }
+
+    if(result == CURLE_OK && !found_certs) {
+      infof(data,
+            "schannel: did not add any certificates from CA file\n");
+    }
+  }while(0);
```

Comment:
See above, add space. And why is this whole block inside a do { ... } while(0); loop? Since I do not see any continue, I don't see the logic behind that. Please elaborate.